bittide/src/Bittide/SwitchDemoProcessingElement.hs

```diff
@@ -0,0 +1,128 @@
+  (linkOut, buffer)
+ where
+  readStartLocked = regEn maxBound (peState .==. pure Idle) readStart
+  readCyclesLocked = regEn maxBound (peState .==. pure Idle) readCycles
+  writeStartLocked = regEn maxBound (peState .==. pure Idle) writeStart
+  writeCyclesLocked = regEn maxBound (peState .==. pure Idle) writeCycles
+
+  localData :: Signal dom (Vec 3 (BitVector 64))
+  localData = bundle (unbundle dnaVec :< (pack <$> localCounter))
+   where
+    dnaVec :: Signal dom (Vec 2 (BitVector 64))
+    dnaVec = bitCoerce . zeroExtend <$> dnaLocked
+    dnaLocked = fromMaybe 0xBAAB_BAAB_BAAB_BAAB_BAAB_BAAB <$> maybeDna
+
+  linkOut = stateToLinkOutput <$> peState <*> buffer <*> localData
+
+  stateToLinkOutput ::
+    SimplePeState nodes ->
+    Vec (nodes * 3) (BitVector 64) ->
+    Vec 3 (BitVector 64) ->
+    BitVector 64
+  stateToLinkOutput state buf locData =
+    case state of
+      Write i
+        | i < 3 -> locData !! i
+        | otherwise -> buf !! i
+      _ -> 0xABBA_ABBA_ABBA_ABBA
+
+  -- \| The buffer stores all the incomming bittide data. For the Bittide Switch demo,
+  -- each FPGA sends its DNA and local clock cycle counter, along with all received data.
+  -- The last FPGA will therefore receive all DNAs and local clock cycle counters.
+  -- Order of datapacket: DNA 32 msbs, DNA 64 lsbs, local counter.
+  -- Order of packets: Current FPGA first, followed by the one before. Comparing all
+  -- buffers would look like a moving window.
+  buffer :: Signal dom (Vec (nodes * 3) (BitVector 64))
+  buffer = bundle $ regEn initVal <$> enableVec <*> linkInVec
+   where
+    initVal = 0xABBA_ABBA_ABBA_ABBA
+    linkInVec = repeat linkIn
+
+    enableVec :: Vec (nodes * 3) (Signal dom Bool)
+    enableVec = unbundle $ go <$> peState
+     where
+      go :: SimplePeState nodes -> Vec (nodes * 3) Bool
+      go (Read x) = (== x) <$> indicesI
+      go _ = repeat False
+
+  peState =
+    register
+      Idle
+      ( update
+          <$> localCounter
+          <*> readStartLocked
+          <*> readCyclesLocked
+          <*> writeStartLocked
+          <*> writeCyclesLocked
+          <*> peState
+      )
+   where
+    update ::
+      -- \| Local clock cycle counter
+      Unsigned 64 ->
+      -- \| When to read from the crossbar link
+      Unsigned 64 ->
+      -- \| How many cycles to read from crossbar link
+      Index (nodes * 3) ->
+      -- \| When to write to the crossbar link
+      Unsigned 64 ->
+      -- \| How many cycles to read from crossbar lin
+      Index (nodes * 3) ->
+      SimplePeState nodes ->
+      SimplePeState nodes
+    update cntr rs rc ws wc state =
+      case state of
+        Idle
+          | cntr == rs -> Read 0
+          | cntr == ws -> Write 0
+          | otherwise -> Idle
+        Read x
+          | x == (rc - 1) -> Idle
+          | otherwise -> Read (satSucc SatBound x)
+        Write x
+          | x == (wc - 1) -> Idle
+          | otherwise -> Write (satSucc SatBound x)
```

Comment:
Could we refactor this to:

```suggestion
    update cntr rs rc ws wc state =
      case state of
        Idle -> nextState
        Read x
          | x == (rc - 1) -> nextState
          | otherwise -> Read (satSucc SatBound x)
        Write x
          | x == (wc - 1) -> nextState
          | otherwise -> Write (satSucc SatBound x)
     where
      nextState
        | cntr == ws -> Write 0
        | cntr == rs -> Read 0
        | otherwise -> Idle
```

That way we can do back-to-back read/write, which makes testing easier.